<a href="https://colab.research.google.com/github/stephiioo/datasci_2_manipulation/blob/main/HHA_507_datasci_manipulation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

#1. Data Cleaning with Pandas

## Loading the csv dataset

In [ ]:
csv1 = pd.read_csv ('https://raw.githubusercontent.com/hantswilliams/HHA_507_2023/main/WK2/data/healthcare_data_cleaning.csv')
csv1

Patient Age  Gender  City of Residence State of Residence  \
0               45   Other         West Brian           Kentucky   
1               25  Female   East Jocelynfurt           Colorado   
2               51   Other  South Lindseyland               Ohio   
3               18  Female         Taylorfort       Pennsylvania   
4               48  Female          Birdmouth            Montana   
...            ...     ...                ...                ...   
104995          74   Other       West Anthony           Oklahoma   
104996          37    Male          Mckeetown               Utah   
104997          51  Female   Lake Lisachester           Nebraska   
104998     missing  Female       Heidiborough            missing   
104999          45  Female      South Jeffrey      Massachusetts   

       Has Insurance Visited Last Month Payment Method Preferred Doctor  \
0                Yes                Yes      Insurance     Dr. Williams   
1                 No                Yes           Card      Dr. Johnson   
2                Yes                 No           Cash     Dr. Williams   
3                Yes            missing           Cash     Dr. Williams   
4                 No                 No      Insurance        Dr. Brown   
...              ...                ...            ...              ...   
104995           Yes                 No           Card     Dr. Williams   
104996           Yes                 No           Cash        Dr. Jones   
104997           Yes                Yes      Insurance     Dr. Williams   
104998           Yes                 No           Cash        Dr. Smith   
104999            No                Yes           Cash        Dr. Brown   

       Disease Diagnosed Medication Prescribed Type of Appointment  \
0                    Flu                 Med_E             General   
1                    Flu                 Med_B          Specialist   
2               Covid-19                 Med_D             General   
3                    Flu                 Med_A             General   
4                   None                 Med_C          Specialist   
...                  ...                   ...                 ...   
104995          Covid-19                 Med_E          Specialist   
104996          Covid-19                 Med_C             General   
104997           Allergy                 Med_D           Follow-Up   
104998          Covid-19                 Med_A           Follow-Up   
104999           Allergy                 Med_D             General   

       Average Heart Rate Average BP Height (in cm) Weight (in kg)  \
0                      66        111            183             70   
1                      59    missing            174             58   
2                      79        119            161             56   
3                      99        115            171             52   
4                      93         89            186             87   
...                   ...        ...            ...            ...   
104995                 65        112            169             61   
104996            missing        111            165             67   
104997                 89        118            179             71   
104998                 74        123            167             75   
104999                 77        140            167             62   

           Payment Due ($) Last Visit (days ago) Visit Duration (mins)  \
0        65.45721578126224                   193                    92   
1       430.68365678679174                   195               missing   
2        315.0709305262176               missing                    37   
3        320.2998987723972                   175                   108   
4        264.2147372473319                   188                   113   
...                    ...                   ...                   ...   
104995  309.16444606680665                   183                    10   
104996         

## Identifying and handling missing values in the dataset

In [ ]:
# check for missing values and count them
missing_values = csv1.isna().sum()

# calculate the percentage of missing values
percentage_missing = (missing_values / len(csv1)) * 100

# Display of the results
print("Missing Values:")
print(missing_values)
print("\nPercent of Missing Values:")
print(percentage_missing)

Missing Values:
Patient Age              0
Gender                   0
City of Residence        0
State of Residence       0
Has Insurance            0
Visited Last Month       0
Payment Method           0
Preferred Doctor         0
Disease Diagnosed        0
Medication Prescribed    0
Type of Appointment      0
Average Heart Rate       0
Average BP               0
Height (in cm)           0
Weight (in kg)           0
Payment Due ($)          0
Last Visit (days ago)    0
Visit Duration (mins)    0
Number of Tests          0
Prescription Cost ($)    0
dtype: int64

Percent of Missing Values:
Patient Age              0.0
Gender                   0.0
City of Residence        0.0
State of Residence       0.0
Has Insurance            0.0
Visited Last Month       0.0
Payment Method           0.0
Preferred Doctor         0.0
Disease Diagnosed        0.0
Medication Prescribed    0.0
Type of Appointment      0.0
Average Heart Rate       0.0
Average BP               0.0
Height (in cm)           0

##Removing any duplicate rows and columns if they exist

In [ ]:
# Remove duplicate rows
csv1_no_duplicates = csv1.drop_duplicates()

# Check for duplicate columns
duplicate_columns = csv1_no_duplicates.columns[csv1_no_duplicates.columns.duplicated()]

# Remove duplicate columns
csv1_no_duplicate_columns = csv1_no_duplicates[csv1_no_duplicates.columns[~csv1_no_duplicates.columns.duplicated()]]

##Cleaning column names

In [ ]:
# Clean the column names
csv1.columns = csv1.columns.str.lower()
csv1.columns = csv1.columns.str.strip()
csv1.columns = csv1.columns.str.replace(' ', '_')
csv1.columns = csv1.columns.str.replace('[^a-zA-Z0-9_]', '', regex=True)

# Print the updated column names
print(csv1.columns)

Index(['patient_age', 'gender', 'city_of_residence', 'state_of_residence',
       'has_insurance', 'visited_last_month', 'payment_method',
       'preferred_doctor', 'disease_diagnosed', 'medication_prescribed',
       'type_of_appointment', 'average_heart_rate', 'average_bp',
       'height_in_cm', 'weight_in_kg', 'payment_due_', 'last_visit_days_ago',
       'visit_duration_mins', 'number_of_tests', 'prescription_cost_'],
      dtype='object')


#2. Data Transformation

##Creating new columns based on existing ones

In [ ]:
#Check the data type of the 'patient_age' column
print(csv1['patient_age'].dtype)

# Convert the 'patient_age' column to numeric
csv1['patient_age'] = pd.to_numeric(csv1['patient_age'], errors = 'coerce')

# Check the data type of the 'patient_age' column after conversion
print(csv1['patient_age'].dtype)

# Define the threshold for geriatric adults
threshold_65 = 65

# Create new column "Age_Group_Status" based on whether the patients are geriatric adults or not based on threshold
csv1['Age_Group_Status'] = csv1['patient_age'].apply(lambda x: 'Geriatric' if x >= threshold_65 else 'Not_Geriatric')

#Print the dataframe with the new 'Age_Group_Status' column
print(csv1)

float64
float64
        patient_age  gender  city_of_residence state_of_residence  \
0              45.0   Other         West Brian           Kentucky   
1              25.0  Female   East Jocelynfurt           Colorado   
2              51.0   Other  South Lindseyland               Ohio   
3              18.0  Female         Taylorfort       Pennsylvania   
4              48.0  Female          Birdmouth            Montana   
...             ...     ...                ...                ...   
104995         74.0   Other       West Anthony           Oklahoma   
104996         37.0    Male          Mckeetown               Utah   
104997         51.0  Female   Lake Lisachester           Nebraska   
104998          NaN  Female       Heidiborough            missing   
104999         45.0  Female      South Jeffrey      Massachusetts   

       has_insurance visited_last_month payment_method preferred_doctor  \
0                Yes                Yes      Insurance     Dr. Williams   
1    

##Aggregating data using groupby and compute summary statistics

In [ ]:
# Define the numeric columns for which to calculate the mean
numeric_columns = ['patient_age', 'average_heart_rate', 'average_bp', 'height_in_cm', 'weight_in_kg', 'visit_duration_mins', 'number_of_tests', 'prescription_cost_']

# Group the data by 'gender' and calculate the mean for the numeric columns
grouped_data = csv1.groupby('gender')
summary_stats = grouped_data[numeric_columns].mean()

# Reset the index
summary_stats.reset_index(inplace=True)

# Print the summary statistics
print(summary_stats)


    gender  patient_age
0   Female    44.692845
1     Male    44.641288
2    Other    44.760662
3  missing    44.769860


<ipython-input-22-1583ab057722>:6: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  summary_stats = grouped_data[numeric_columns].mean()


##Cross-tabulation using multidimentional analysis

In [23]:
# Create a cross-tabulation table for 'Gender' and 'Age_Group'
cross_tab = pd.crosstab(csv1['gender'], csv1['patient_age'])

# Display the cross-tabulation table
cross_tab

patient_age  18.0   19.0   20.0   21.0   22.0   23.0   24.0   25.0   26.0   \
gender                                                                       
Female        1325    218    237    265    280    291    320    361    412   
Male          1403    186    212    242    246    285    315    373    386   
Other         1313    208    231    257    302    263    296    371    399   
missing        214     28     31     47     58     48     48     53     74   

patient_age  27.0   ...  91.0   92.0   93.0   94.0   95.0   96.0   97.0   \
gender              ...                                                    
Female         411  ...      3      7      6      5      3      4      3   
Male           425  ...      4      7      8      5      8      1      2   
Other          434  ...      3      4      5      4      8      5      2   
missing         67  ...      0      2      2      1      0      0      1   

patient_age  98.0   99.0   100.0  
gender                            
Female           2      2      0  
Male             2      3      8  
Other            0      3      2  
missing          1      0      0  

[4 rows x 83 columns]

##Introduction to Alternative Libraries

In [33]:
# Import polars
import polars as pl

#Import time
import time

# Define the URL of CSV dataset
csv1_url = 'https://raw.githubusercontent.com/hantswilliams/HHA_507_2023/main/WK2/data/healthcare_data_cleaning.csv'

# Measure the start time
start_time = time.time()

# Load dataset using polars
csv1 = pl.read_csv(csv1_url)

# Measure the end time
end_time = time.time()

# Calculate the load time
load_time = end_time - start_time

# Display the start and end load times as well as the total load times
print(f"Start Time: {start_time}")
print(f"End Time: {end_time}")
print(f"Load Time: {load_time: .2f} seconds")

# Display the first few rows of the loaded dataset for verification
print("\nLoaded Dataset:")
print(csv1.head())

Start Time: 1694566385.2209425
End Time: 1694566385.490973
Load Time:  0.27 seconds

Loaded Dataset:
shape: (5, 20)
┌───────────┬────────┬────────────┬────────────┬───┬────────────┬────────────┬────────────┬────────────┐
│ Patient   ┆ Gender ┆ City of    ┆ State of   ┆ … ┆ Last Visit ┆ Visit      ┆ Number of  ┆ Prescripti │
│ Age       ┆ ---    ┆ Residence  ┆ Residence  ┆   ┆ (days ago) ┆ Duration   ┆ Tests      ┆ on Cost    │
│ ---       ┆ str    ┆ ---        ┆ ---        ┆   ┆ ---        ┆ (mins)     ┆ ---        ┆ ($)        │
│ str       ┆        ┆ str        ┆ str        ┆   ┆ str        ┆ ---        ┆ str        ┆ ---        │
│           ┆        ┆            ┆            ┆   ┆            ┆ str        ┆            ┆ str        │
╞═══════════╪════════╪════════════╪════════════╪═══╪════════════╪════════════╪════════════╪════════════╡
│ 45        ┆ Other  ┆ West Brian ┆ Kentucky   ┆ … ┆ 193        ┆ 92         ┆ 5          ┆ 15.7175173 │
│           ┆        ┆            ┆         

In [36]:
# Import polars
import polars as pl

#Import time
import time

# Define the URL of CSV dataset
csv2_url = 'https://www.kaggle.com/datasets/fedesoriano/heart-failure-prediction'

# Measure the start time
start_time = time.time()

# Load dataset using polars
csv2 = pl.read_csv(csv2_url)

# Measure the end time
end_time = time.time()

# Calculate the load time
load_time = end_time - start_time

# Display the start and end load times as well as the total load times
print(f"Start Time: {start_time}")
print(f"End Time: {end_time}")
print(f"Load Time: {load_time: .2f} seconds")

# Display the first few rows of the loaded dataset for verification
print("\nLoaded Dataset:")
print(csv2.head())

Start Time: 1694566889.1491487
End Time: 1694566889.346335
Load Time:  0.20 seconds

Loaded Dataset:
shape: (5, 1)
┌──────────────────┐
│ <!DOCTYPE html>  │
│ ---              │
│ str              │
╞══════════════════╡
│ null             │
│ <!DOCTYPE html>  │
│ <html lang="en"> │
│ null             │
│ <head>           │
└──────────────────┘


##Comparing load times and making observations for datasets using polars

In this analysis, we will compare the load times for two datasets using the polars library in Google Colab with the interest of measuring the time that it takes to load each dataset to assess their loading performance.

### Dataset Information

1. **Dataset 1:** This dataset contains patient health information including patient demographics, medical diagnosis, health statistics and much more

2. **Dataset 2:** This dataset contains patient health information that is used to predict their chances of heart failure

### Results

Here are the load times for each dataset:

- **Dataset 1 Load Time:** 0.27 seconds
- **Dataset 2 Load Time:** 0.20 seconds

Dataset 2 loaded significantly faster than Dataset 1 which suggests that it may have a more efficient data structure or a smaller file size.